Framework: Replace element-closest with registered vendor script

aduth opened the pull request:
Closes #7159

This pull request seeks to eliminate `element-closest` dependency and imports from individual packages, instead replacing with a standard polyfill approach via server-side script enqueue.

Potential future improvements include:

- Improved association between script handle as being a polyfill (avoid `wp_add_inline_script` per dependent)
- Avoid polyfill dependencies needing to be kept in sync with current state of codebase using / not using (maybe always load)

**Testing instructions:**

Verify there are no regressions in `Element#closest` behaviors, notably in IE11 where it is not supported by default. For example, verify that inserting a quote block and ArrowDown'ing through its citation and into the addition of a default block (verifies usage of `Element#closest` in WritingFlow's use of `isInSameBlock`).

youknowriad commented:
I wonder if we should just do it for `wp-polyfill-ecmascript` (adding the dependency). This would avoid us to keep track of the exact usage of this feature per script which is very hard to accomplish anyway.

aduth commented:
@youknowriad Yes, I wasn't too sure how we might want to accomplish the "load always" behavior for `element-closest`. There's a bit of additional context in #7159 and #7033. With your suggestion, we'd expand what's implied by adding a dependency on `wp-polyfill-ecmascript` from meaning "latest JavaScript features" to also including "modern browser behaviors". These often collide anyways in our usage, and combining them would make for a simpler developer experience, but they do hold separate meanings which could be worth distinguishing.

Maybe a topic for tomorrow's chat? 🙂

aduth commented:
Put another way, ECMAScript and WHATWG/W3 standards are separate, though often used interchangeably as we use modern features and for polyfilling. Should we or shouldn't we abandon attempts to keep them distinct?

youknowriad commented:
`core-js` which is basically what `babel-polyfill` uses is not only about `ECMAScript` but it also includes some `WHATWG/W3` standards. I'm not really certain if `babel-polyfill` picks only the `ECMAScript` features thought or just enqueues the whole thing.

aduth commented:
>core-js which is basically what babel-polyfill uses is not only about ECMAScript but it also includes some WHATWG/W3 standards.

Is this true? Could you name some such features?

Genuinely curious because I'd commented as such that my assumption is that it's only for ECMAScript:

https://github.com/WordPress/gutenberg/pull/9171#issuecomment-419536905

aduth commented:
Oh I see now from their documentation they have a separate section for web standards:

https://github.com/zloirock/core-js#features

Then it becomes a question of: Are we aligned to `core-js`, or to `babel-polyfill`? And does `babel-polyfill` make the same commitments?

aduth committed:
Framework: Replace element-closest with registered vendor script

aduth committed:
Scripts: Add support for license ignore

aduth committed:
Framework: Re-pin to newer version of JSDOM

Newer version is SemVer-compatible with jest-environment-jsdom and includes support for Element#closest out of the box.

https://github.com/jsdom/jsdom/issues/1555

Required to ignore license validation for abab because while it is a compatible license (BSD 3-clause), the package's `license` field is malformed for parse.

https://github.com/jsdom/abab/blob/4327de3aae348710094d9f3c1f0c1477d9feb865/package.json#L26
https://github.com/jsdom/abab/blob/master/LICENSE.md

aduth commented:
Rebased to account for the now-merged #9794, which simplifies the implementation here a fair bit.

I also discovered that newer versions of JSDOM include support for `Element#closest` (https://github.com/jsdom/jsdom/issues/1555) such that it wouldn't be necessary to explicitly provide this in `jest-preset-default`. See f6916bd.

In updating our pinned version, however, I fought with the license checker flagging `abab`. I think we've seen this on a few occasions before. Ultimately I decided to add an `--ignore` option to the license checker, to be used for these sorts of cases where we can manually vet the license for a package as being compatible. The `abab` package has a compatible license (W3C BSD 3-Clause), but because [it is specified in its `package.json` in a non-parseable format](
https://github.com/jsdom/abab/blob/4327de3aae348710094d9f3c1f0c1477d9feb865/package.json#L26), it is marked as being invalid. See 674e70f.

aduth commented:
cc @pento re: changes to `wp-scripts check-licenses`

aduth committed:
Testing: Add dummy testURL for Jest config

Avoids issues with localStorage in JSDOM 11.12

Can be removed when running Jest 23.5.0 or newer

See: https://github.com/facebook/jest/issues/6766
See: https://github.com/facebook/jest/pull/6792

pento committed:
Licenses: Add BSD-3-Clause W3C variant as an allowed license

Also, check if the license defined in package.json is telling us to look in the license file, instead.

pento commented:
Licenses are the worst. I mean, I get why they don't put an identifier in the `license` field, because there isn't a SPDX identifier for the W3C variant of the BSD 3-Clause license. But they could just make it up.

I like the `--ignore` option, but I hope we never have to use it. 65bf762 adds `BSD-3-Clause-W3C` as a GPL2 compatible license, and adds a check for if the `license` field is telling us to look in a license file, instead.

aduth commented on packages/scripts/scripts/check-licenses.js:
```diff
@@ -193,6 +225,11 @@ modules.forEach( ( path ) => {
 
+	// Check if the license we've detected is telling us to look in the license file, instead.
+	if ( licenseType && licenseFiles.find( ( licenseFile ) => licenseType.indexOf( licenseFile ) >= 0 ) ) {
+		licenseType = undefined;
```
What if the `license` property is pointing a file describing a truly incompatible license? Wouldn't this wrongly allow the license to be considered valid? I expect this to be even more common for non-standard licenses.

aduth replied:
Oh! I missed the code immediately following this, where we try to infer the license type from the presence of a file.

pento replied:
🙂

aduth commented:
Rebased to resolve conflicts.

At this point just need a 👍

brandonpayton reviewed (approved):
I tested this in IE11, and everything appears to work as expected. I also checked this in Chrome and observed that the native version of `HTMLElement.prototype.closest` was intact.

👍  LGTM

brandonpayton commented on packages/scripts/scripts/check-licenses.js:
```diff
@@ -193,6 +225,11 @@ modules.forEach( ( path ) => {
 	let licenseType = typeof license === 'object' ? license.type : license;
 
+	// Check if the license we've detected is telling us to look in the license file, instead.
+	if ( licenseType && licenseFiles.find( ( licenseFile ) => licenseType.indexOf( licenseFile ) >= 0 ) ) {
```
nit: This is a good place to use `String#includes` which reads a bit better than an index comparison.

mcsf reviewed (approved):
Good change. :)

aduth committed:
Package: Scripts: Use String#includes for improved code clarity

aduth committed:
Package: Scripts: Add CHANGELOG notes for improved license file detection
